File: Source/Aladdin.CAPI.ANSI.Java/src/aladdin/capi/ansi/keyx/rsa/oaep/Encoding.java
package aladdin.capi.ansi.keyx.rsa.oaep;
import aladdin.capi.*;
import aladdin.util.*; 
import java.io.*;
import java.util.Arrays;

///////////////////////////////////////////////////////////////////////////
// Кодирование EME OAEP
///////////////////////////////////////////////////////////////////////////
public class Encoding
{
    // закодировать данные
    public static byte[] encode(Hash hashAlgorithm, 
        PRF maskAlgorithm, byte[] label, IRand rand, byte[] data, int k) throws IOException
    {
        // определить размер хэш-значения
        int hLen = hashAlgorithm.hashSize(); byte[] encoded = new byte[k]; 

        // проверить размер данных
        if (data.length > k - 2 * hLen - 2) throw new IllegalArgumentException();

        // вычислить хэш-значение от метки
        byte[] lHash = hashAlgorithm.hashData(label, 0, label.length);

        // создать нулевое заполнение
        byte[] PS = new byte[k - data.length - 2 * hLen - 2]; encoded[0] = 0x00; 

        // создать блок данных
        byte[] DB = Array.concat(lHash, PS, new byte[] {0x01}, data); 

        // сгенерировать случайное значение
        byte[] seed = new byte[hLen]; rand.generate(seed, 0, hLen); 
        
        // вычислить маску
        maskAlgorithm.generate(seed, null, encoded, 1 + hLen, k - hLen - 1);  
        
        // сложить блок данных с маской
        for (int i = 0; i < k - hLen - 1; i++) DB[i] ^= encoded[1 + hLen + i]; 

        // скопировать замаскированный блок данных
        System.arraycopy(DB, 0, encoded, 1 + hLen, k - hLen - 1); 
        
        // вычислить маску
        maskAlgorithm.generate(DB, null, encoded, 1, hLen);   
        
        // замаскировать случайное значение
        for (int i = 0; i < hLen; i++) encoded[1 + i] ^= seed[i]; return encoded; 
    }
    // раскодировать данные
    public static byte[] decode(Hash hashAlgorithm, 
        PRF maskAlgorithm, byte[] label, byte[] encoded) throws IOException
    {
        // проверить значение первого байта
        if (encoded[0] != 0) throw new IOException(); int k = encoded.length;

        // определить размер хэш-значения
        int hLen = hashAlgorithm.hashSize(); byte[] seed = new byte[hLen];

        // вычислить хэш-значение 
        byte[] lHash = hashAlgorithm.hashData(label, 0, label.length);

        // извлечь замаскированный блок данных
        byte[] DB = new byte[k - hLen - 1]; System.arraycopy(encoded, 1 + hLen, DB, 0, k - hLen - 1);
        
        // вычислить маску
        maskAlgorithm.generate(DB, null, seed, 0, hLen); 
        
        // вычислить случайное значение
        for (int i = 0; i < hLen; i++) seed[i] ^= encoded[1 + i]; 

        // вычислить маску
        maskAlgorithm.generate(seed, null, DB, 0, k - hLen - 1);  
        
        // вычислить блок данных
        for (int i = 0; i < k - hLen - 1; i++) DB[i] ^= encoded[1 + hLen + i]; 

        // проверить совпадение хэш-значения
        if (!Array.equals(DB, 0, lHash, 0, hLen)) throw new IOException();

        // пропустить нулевые байты
        int j; for (j = hLen; j < k - hLen - 1 && DB[j] == 0; j++) {} 

        // проверить корректность данных
        if (j == k - hLen - 1 || DB[j] != 0x01) throw new IOException();

        // выделить память для результата
        byte[] data = new byte[k - hLen - j - 2]; 

        // извлечь данные
        System.arraycopy(DB, j + 1, data, 0, data.length); return data; 
    }
}

File: Source/Aladdin.CAPI.ANSI/CAPI/ANSI/Sign/RSA/PSS/Encoding.cs
using System;
using System.IO;

namespace Aladdin.CAPI.ANSI.Sign.RSA.PSS
{
    ///////////////////////////////////////////////////////////////////////////
    // Кодирование EMSA PSS
    ///////////////////////////////////////////////////////////////////////////
    public static class Encoding
    {
	    public static byte[] Encode(IRand rand, CAPI.Hash hashAlgorithm, 
            PRF maskAlgorithm, byte trailerField, int emBits, int sLen, byte[] hash)
	    {
            // сгенерировать случайные данные
            byte[] salt = new byte[sLen]; rand.Generate(salt, 0, sLen); 
                
            // объединить данные последующего хэширования
            byte[] M = Arrays.Concat(new byte[8], hash, salt); 

            // вычислить хэш-значение
            byte[] H = hashAlgorithm.HashData(M, 0, M.Length); 

		    // определить размер хэш-значения
		    int hLen = H.Length; int emLen = (emBits + 7) / 8;

            // проверить корректность параметров
            if (sLen + hLen + 2 > emLen) throw new InvalidDataException(); 

		    // создать нулевое дополнение
		    byte[] PS = new byte[emLen - sLen - hLen - 2]; byte[] encoded = new byte[emLen]; 

		    // создать блок данных 
		    byte[] DB = Arrays.Concat(PS, new byte[] {0x01}, salt);   

            // вычислить маску
	        maskAlgorithm.Generate(H, null, encoded, 0, emLen - hLen - 1); 

    	    // скопировать хэш-значение
		    Array.Copy(H, 0, encoded, emLen - hLen - 1, hLen); 

		    // замаскировать блок данных
		    for (int i = 0; i < emLen - hLen - 1; i++) encoded[i] ^= DB[i]; 

		    // обнулить неиспользуемые биты
		    if ((emBits % 8) != 0) encoded[0] &= (byte)((1 << (emBits % 8)) - 1);  

		    // установить значение завершителя
		    encoded[emLen - 1] = trailerField; return encoded; 
	    }
	    public static void Decode(CAPI.Hash hashAlgorithm, PRF maskAlgorithm, 
		    byte trailerField, byte[] encoded, int emBits, int sLen, byte[] hash)
	    {
		    // определить размер хэш-значения
		    int hLen = hashAlgorithm.HashSize; int emLen = (emBits + 7) / 8;

		    // проверить корректность неиспользуемых битов
		    if ((emBits % 8) != 0 && (encoded[0] & ~((1 << (emBits % 8)) - 1)) != 0)
            {
                // выбросить исключение
                throw new InvalidDataException(); 
            }
		    // проверить наличие завершителя
		    if (encoded[emLen - 1] != trailerField) throw new InvalidDataException();

            // выделить память для salt-значения и хэш-значения
            byte[] salt = new byte[sLen]; byte[] H = new byte[hLen]; 

		    // извлечь хэш-значение 
		    Array.Copy(encoded, emLen - hLen - 1, H, 0, hLen); 
 
            // выделить буфе для маски
            byte[] DB = new byte[emLen - hLen - 1]; 
            
	        // вычислить маску
	        maskAlgorithm.Generate(H, null, DB, 0, emLen - hLen - 1); 
            
		    // вычислить блок данных
		    for (int i = 0; i < emLen - hLen - 1; i++) DB[i] ^= encoded[i]; 

		    // обнулить неиспользуемые биты
		    if ((emBits % 8) != 0) DB[0] &= (byte)((1 << (emBits % 8)) - 1); 

		    // проверить наличие нулевых байтов
		    for (int i = 0; i < emLen - hLen - sLen - 2; i++) 
            {
		        // проверить наличие нулевых байтов
                if (DB[i] != 0) throw new InvalidDataException();
            }
		    // проверить наличие разделителя
		    if (DB[emLen - hLen - sLen - 2] != 0x01) throw new InvalidDataException();

		    // извлечь salt-значение
		    Array.Copy(DB, emLen - sLen - hLen - 1, salt, 0, sLen); 

            // объединить данные последующего хэширования
            byte[] M = Arrays.Concat(new byte[8], hash, salt); 
                
            // вычислить хэш-значение
            byte[] check = hashAlgorithm.HashData(M, 0, M.Length); 

            // сравнить хэш-значения
            if (!Arrays.Equals(H, check)) throw new SignatureException(); 
	    }
    }
}
